envoy/upstream/load_balancer.h

```diff
@@ -130,13 +164,33 @@ class LoadBalancer {
 public:
   virtual ~LoadBalancer() = default;
 
+  /*
+   * This is a convenience wrapper function for code which does not yet support
+   * asynchronous host selection. It cancels any asynchronous lookup and treats
+   * it as host selection failure.
+   */
+  static HostConstSharedPtr
+  onlyAllowSynchronousHostSelection(HostSelectionResponse host_selection) {
+    if (host_selection.cancelable) {
+      // Async host selection not handled yet. Treat this as host selection
+      // failure.
+      host_selection.cancelable->cancel();
+    }
+    return std::move(host_selection.host);
+  }
+
   /**
    * Ask the load balancer for the next host to use depending on the underlying LB algorithm.
    * @param context supplies the load balancer context. Not all load balancers make use of all
    *        context information. Load balancers should be written to assume that context information
    *        is missing and use sensible defaults.
+   * @return a HostSelectionResponse either containing a host, or AsyncHostSelection handle.
+   *
+   * Please note that asynchronous host selection is not yet fully supported in
+   * Envoy. It works for HTTP load balancing (with the notable exclusion of the
+   * subset load balancer) but not for TCP proxy or other load balancers.
    */
-  virtual HostConstSharedPtr chooseHost(LoadBalancerContext* context) PURE;
+  virtual HostSelectionResponse chooseHost(LoadBalancerContext* context) PURE;
```

Comment:
I could do a pure API change here with no one returning anything but the host, and every call site only allowing async options

Reply:
FIWW this one would still be a giant change, but a giant more-clearly-no-op change which might be worth it?  your call

Reply:
This definitely sounds appetizing, I think. Yeah, after looking at this in more details, I think this would be a good approach.

Reply:
Ok one request, before I go and do that can you do a pass on the API (just load_balancer.h changes) and make sure you're ok with it at a high level?
basically 
chooseHost goes from returning a host to 
1) returning a host or
2) returning a cancelable handle, promising to call onAsyncHostSelected if cancelable->cancel isn't called

If you want anything renamed I'd rather do it in the original PR than split them out, do a rename, then have a merge from hell :-)

Reply:
actually nm, I think I can do this easily enough it won't be a pain